src/goto-instrument/contracts/dynamic-frames/dfcc.cpp

```diff
@@ -0,0 +1,497 @@
+
+Module: Dynamic frame condition checking
+
+Author: Remi Delmas, delmarsd@amazon.com
+
+\*******************************************************************/
+
+#include "dfcc.h"
+
+#include <util/config.h>
+#include <util/expr_util.h>
+#include <util/format_expr.h>
+#include <util/format_type.h>
+#include <util/fresh_symbol.h>
+#include <util/mathematical_expr.h>
+#include <util/mathematical_types.h>
+#include <util/namespace.h>
+#include <util/optional.h>
+#include <util/pointer_expr.h>
+#include <util/pointer_offset_size.h>
+#include <util/pointer_predicates.h>
+#include <util/prefix.h>
+#include <util/std_expr.h>
+
+#include <goto-programs/goto_convert_functions.h>
+#include <goto-programs/goto_functions.h>
+#include <goto-programs/goto_inline.h>
+#include <goto-programs/goto_model.h>
+#include <goto-programs/initialize_goto_model.h>
+#include <goto-programs/link_to_library.h>
+#include <goto-programs/remove_skip.h>
+#include <goto-programs/remove_unused_functions.h>
+
+#include <ansi-c/ansi_c_entry_point.h>
+#include <ansi-c/c_expr.h>
+#include <ansi-c/c_object_factory_parameters.h>
+#include <ansi-c/cprover_library.h>
+#include <goto-instrument/contracts/cfg_info.h>
+#include <goto-instrument/contracts/instrument_spec_assigns.h>
+#include <goto-instrument/contracts/utils.h>
+#include <goto-instrument/nondet_static.h>
+#include <langapi/language.h>
+#include <langapi/language_file.h>
+#include <langapi/mode.h>
+#include <linking/static_lifetime_init.h>
+
+void dfcc(
+  const optionst &options,
+  goto_modelt &goto_model,
+  const irep_idt &harness_id,
+  const std::set<irep_idt> &to_check,
+  const bool allow_recursive_calls,
+  const std::set<irep_idt> &to_replace,
+  const bool apply_loop_contracts,
+  const std::set<std::string> &to_exclude_from_nondet_static,
+  message_handlert &message_handler)
+{
+  std::map<irep_idt, irep_idt> to_check_map;
+  for(const auto &function_id : to_check)
+    to_check_map.insert({function_id, function_id});
+
+  std::map<irep_idt, irep_idt> to_replace_map;
+  for(const auto &function_id : to_replace)
+    to_replace_map.insert({function_id, function_id});
+
+  dfcct{options, goto_model, message_handler}.transform_goto_model(
+    harness_id,
+    to_check_map,
+    allow_recursive_calls,
+    to_replace_map,
+    apply_loop_contracts,
+    to_exclude_from_nondet_static);
+}
+
+void dfcc(
+  const optionst &options,
+  goto_modelt &goto_model,
+  const irep_idt &harness_id,
+  const std::map<irep_idt, irep_idt> &to_check,
+  const bool allow_recursive_calls,
+  const std::map<irep_idt, irep_idt> &to_replace,
+  const bool apply_loop_contracts,
+  const std::set<std::string> &to_exclude_from_nondet_static,
+  message_handlert &message_handler)
+{
+  dfcct{options, goto_model, message_handler}.transform_goto_model(
+    harness_id,
+    to_check,
+    allow_recursive_calls,
+    to_replace,
+    apply_loop_contracts,
+    to_exclude_from_nondet_static);
```

Comment:
Shouldn't the code of `transform_goto_model` just be inlined in here? Why wrap another layer around it?

Reply:
the dfcct class is an instrumentation detail, the dfcc functions are the ones to use.